### PR TITLE
HV-1328 Add an option to validate class-level constraints only if all property constraints are valid

### PR DESCRIPTION
**Pull request by marko-bekhta**
- https://hibernate.atlassian.net/browse/HV-1328

First commit has a change for what looked like a strange behavior. the value only for the last group was considered....

Any ideas on naming of the "mode" is also welcomed. For now PR is missing the doc.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java [hunk 1]
```diff
@@ -471,7 +471,7 @@ private void validateConstraintsForCurrentGroup(BaseBeanValidationContext<?> val
 
 						for ( Group defaultSequenceMember : groupOfGroups ) {
 							validationSuccessful = validateConstraintsForSingleDefaultGroupElement( validationContext, valueContext, validatedInterfaces, clazz,
-									metaConstraints, defaultSequenceMember );
+									metaConstraints, defaultSequenceMember ) && validationSuccessful;
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java]: Could you open a separate issue for that one? A test case might be in order too as it looks a lot like a bug!

File: engine/src/main/java/org/hibernate/validator/BaseHibernateValidatorConfiguration.java [hunk 2]
```diff
@@ -119,6 +119,16 @@
 	@Incubating
 	String GETTER_PROPERTY_SELECTION_STRATEGY_CLASSNAME = "hibernate.validator.getter_property_selection_strategy";
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/BaseHibernateValidatorConfiguration.java]: typo in commit message: s/thorough/through/

File: engine/src/main/java/org/hibernate/validator/BaseHibernateValidatorConfiguration.java [hunk 3]
```diff
@@ -119,6 +119,16 @@
...
 
+
+	/**
+	 * Property corresponding to the {@link #ignoreClassLevelOnPropertyFail} method.
+	 * Accepts {@code true} or {@code false}. Defaults to {@code false}.
+	 *
+	 * @since 6.1.0
+	 */
+	@Incubating
+	String IGNORE_CLASS_LEVEL_ON_PROPERTY_FAIL = "hibernate.validator.ignore_class_level_on_property_fail";
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/BaseHibernateValidatorConfiguration.java]: I wonder if we should try to find a naming similar to the fail fast mode as it's really similar to it.

Maybe `fail_fast_on_property_violation`?

It's not easy to find a right name for this sort of configuration knob.

Happy to discuss it further (maybe it will be easier on Zulip).

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java [hunk 4]
```diff
@@ -463,15 +463,19 @@ private void validateConstraintsForCurrentGroup(BaseBeanValidationContext<?> val
...
 				while ( defaultGroupSequence.hasNext() ) {
 					for ( GroupWithInheritance groupOfGroups : defaultGroupSequence.next() ) {
 						boolean validationSuccessful = true;
 
 						for ( Group defaultSequenceMember : groupOfGroups ) {
-							validationSuccessful = validateConstraintsForSingleDefaultGroupElement( validationContext, valueContext, validatedInterfaces, clazz,
-									metaConstraints, defaultSequenceMember ) && validationSuccessful;
+							boolean propertyValidationResult = validateConstraintsForSingleDefaultGroupElement( validationContext, valueContext, validatedInterfaces, clazz, propertyMetaConstraints, defaultSequenceMember );
+							validationSuccessful = propertyValidationResult && validationSuccessful;
+							if ( !validationContext.isIgnoreClassLevelOnPropertyFailModeEnabled() || propertyValidationResult ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java]: I'm wondering if having these two calls could slow down things. I wonder if we should have a shortcut if this mode is disabled.

Also wondering if we should test if the set is empty (can't say if it will be a win though but it looks like something we should do - and bench).

In any case, we should test it.

File: engine/src/test/java/org/hibernate/validator/test/internal/engine/groups/sequence/SequenceTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		Validator validator = ValidatorUtil.getValidator();
+
+		Set<ConstraintViolation<Foo>> violations = validator.validate( new Foo( null, "", null ) );
+
+		assertThat( violations ).containsOnlyViolations(
+				violationOf( NotNull.class ).withProperty( "str1" )
+		);
+	}
+
+	interface Group1 extends Group11, Group12, Group13, Group14, Group15, Group16, Group17, Group18, Group19 {
```
Comment (marko-bekhta) [inline engine/src/test/java/org/hibernate/validator/test/internal/engine/groups/sequence/SequenceTest.java]: As that loop in which `validationSuccessful` is set is based on a `Set` it actually doesn't give consistent results. To be able to reproduce it more often I've added a lot of groups, which increase the probability that `true` will override the result in the end.

**Comment by marko-bekhta** [comment]
And here's the benchmark results that I've got from yesterday:
```bash
6.1
Benchmark                                   Mode  Cnt     Score     Error   Units
SimpleValidation.testSimpleBeanValidation  thrpt   20  1595.544 ± 127.297  ops/ms

CascadedValidation.testCascadedValidation                                                  thrpt   20   377.533 ±  6.882  ops/ms
CascadedWithLotsOfItemsAndMoreConstraintsValidation.testCascadedValidationWithLotsOfItems  thrpt   20  1012.516 ± 12.189   ops/s
CascadedWithLotsOfItemsValidation.testCascadedValidationWithLotsOfItems                    thrpt   20  1147.559 ± 13.166   ops/s

current:
Benchmark                                   Mode  Cnt     Score     Error   Units
SimpleValidation.testSimpleBeanValidation  thrpt   20  1628.147 ± 185.557  ops/ms

CascadedValidation.testCascadedValidation                                                  thrpt   20   332.512 ±  5.129  ops/ms
CascadedWithLotsOfItemsAndMoreConstraintsValidation.testCascadedValidationWithLotsOfItems  thrpt   20  1039.538 ± 12.060   ops/s
CascadedWithLotsOfItemsValidation.testCascadedValidationWithLotsOfItems                    thrpt   20  1127.804 ± 13.872   ops/s
```

**Comment by gsmet** [comment]
@marko-bekhta do we have a benchmark with property constraints + class level constraints? Because that would be an interesting one :).

I'm not sure the others are significant in this case. Although we should probably have a closer look to `CascadedValidation.testCascadedValidation`.

**Comment by marko-bekhta** [comment]
jenkins retest this please

**Comment by marko-bekhta** [comment]
I've added another benchmark that has both class and property level constraints. The results are:
**no changes:**
```
Benchmark                                                Mode  Cnt     Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  1452.835 ± 21.071  ops/ms
```
**with changes:**
```
Benchmark                                                Mode  Cnt     Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  1438.992 ± 216.247  ops/ms
```
I also tried to run the cascading benchmarks again. The results are all around the place :roll_eyes: . I'll try to run these on my laptop later today to see how they would behave there.

**Comment by marko-bekhta** [comment]
While looking through the PRs I've realized that I had not added those benchmarks.... So here they are:
```bash
#nochange (current)
Benchmark                                                Mode  Cnt     Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  1694.371 ± 22.571  ops/ms

#chagne (changes in this PR)
Benchmark                                                Mode  Cnt     Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  1518.953 ± 29.393  ops/ms

# update (not in this PR, additional if to check if the option is enabled and if not use the "current" all constraints collections)
Benchmark                                                Mode  Cnt     Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  1679.028 ± 33.595  ops/ms
```

**Comment by gsmet** [comment]
@marko-bekhta so I think it makes it clear we need the additional test. Any chance you could push this to the PR (your report above says it wasn't pushed)?

**Comment by gsmet** [comment]
I need to do a quick release so I'm merging the first commit to master and 6.0 as it looks like a bug to me.

**Comment by marko-bekhta** [comment]
@gsmet yes, it totally make sense to break this into two pieces. I'll look for that additional commit and push it here when I find it :)

**Comment by gsmet** [comment]
@marko-bekhta hey Marko. I tried to contact you by email but I didn't hear from you? Did you get my email from December?

File: documentation/src/main/asciidoc/ch12.asciidoc [hunk 6]
```diff
@@ -855,3 +855,52 @@ should be used as a default way of resolving the name (see how it is used in <<e
 - `org.hibernate.validator.spi.nodenameprovider.JavaBeanProperty` is an interface that holds metadata about a bean property. It
 extends `org.hibernate.validator.spi.nodenameprovider.Property` and provide some additional methods like `Class<?> getDeclaringClass()`
 which returns the class that is the owner of the property.
+
+[[section-fail-fast-on-property-violation]]
+=== Fail fast on property violation mode
+
+Fail fast on property violation mode allows to perform class level constraint validation only if all property level constrints have passed without any violations.
+Enabling such mode should make writing class level constraint validators easier and cleaner, as the user will be working with valid properties.
+Hance there will be no need in any additional `null` checks or data sanity checks.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch12.asciidoc:864]:
```suggestion
Hence there will be no need in any additional `null` checks or data sanity checks.
```

Comment (marko-bekhta) [reply documentation/src/main/asciidoc/ch12.asciidoc]: thanks for noticing the typo :wink: updated the text and rebased the branch as it was quite outdated

**Comment by marko-bekhta** [comment]
I've rebased this PR and tried the benchmarks:
Changes:
```java
Benchmark                                                Mode  Cnt      Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  12765.571 ± 84.695  ops/ms
```
main:
```java
Benchmark                                                Mode  Cnt      Score    Error   Units
SimpleClassPropertyValidation.testSimpleBeanValidation  thrpt   20  12672.389 ± 11.129  ops/ms
```

The longer a PR is opened, the better the numbers get (https://github.com/hibernate/hibernate-validator/pull/1031#issuecomment-521964665) 😆 😄 

Jokes aside, I'm not sure whether we should merge it or just close it, as the community has shown little interest in it all this time. 

I'll let you decide what to do about it @yrodiere

**Comment by yrodiere** [comment]
I see the value in this feature, though I'm not sure it's a silver bullet. I suspect that anyone doing a `if ( bean instanceof ) { ... } else if ... ` in their class-level validation constraint will discover the hard way that subclass property constraints haven't been validated at that point.

Anyway... I'd vote for marking this feature as incubating, merging it and moving on :)

**Comment by marko-bekhta** [comment]
> Anyway... Assuming the effect is minimal when the feature isn't enabled, I'd vote for marking this feature as incubating, merging it and moving on :)

thanks 😃. It's marked as incubating. I'll merge it into 9.0, and we can see if we get any feedback on it. 🤞🏻
